src/v/kafka/server/audit_log_manager.cc

```diff
@@ -0,0 +1,569 @@
+      config::shard_local_cfg().audit_max_queue_elements_per_shard.bind())
+  , _audit_event_types(
+      config::shard_local_cfg().audit_enabled_event_types.bind())
+  , _controller(controller)
+  , _config(client_config) {
+    if (ss::this_shard_id() == audit_client::shard_id) {
+        _sink = std::make_unique<audit_sink>(this, controller, client_config);
+    }
+
+    _drain_timer.set_callback([this] {
+        ssx::spawn_with_gate(_gate, [this]() {
+            return drain()
+              .handle_exception([](std::exception_ptr e) {
+                  vlog(
+                    klog.warn, "Exception in audit_log_manager fiber: {}", e);
+              })
+              .finally(
+                [this] { _drain_timer.arm(_queue_drain_interval_ms()); });
+        });
+    });
+    set_enabled_events();
+}
+
+audit_log_manager::~audit_log_manager() = default;
+
+bool audit_log_manager::is_audit_event_enabled(
+  audit_event_type event_type) const {
+    using enum_t = std::underlying_type_t<audit_event_type>;
+    return (_enabled_event_types & enum_t(event_type)) != 0;
+}
+
+ss::future<> audit_log_manager::start() {
+    _auditing_enabled.watch([this] {
+        if (ss::this_shard_id() == audit_client::shard_id) {
+            try {
+                _sink->toggle(_auditing_enabled());
+            } catch (const ss::gate_closed_exception&) {
+                vlog(
+                  klog.debug, "Failed to toggle auditing state, shutting down");
+            } catch (...) {
+                vlog(
+                  klog.error,
+                  "Failed to toggle auditing state: {}",
+                  std::current_exception());
+            }
+        }
+    });
+    _audit_event_types.watch([this] { set_enabled_events(); });
+    if (_auditing_enabled()) {
+        _drain_timer.arm(_queue_drain_interval_ms());
+        if (ss::this_shard_id() == audit_client::shard_id) {
+            vlog(klog.info, "Starting audit_log_manager");
+            co_await _sink->start();
+        }
+    }
+}
+
+ss::future<> audit_log_manager::stop() {
+    _drain_timer.cancel();
+    _as.request_abort();
+    if (ss::this_shard_id() == audit_client::shard_id) {
+        vlog(klog.info, "Shutting down audit log manager");
+        co_await _sink->stop();
+    }
+    if (!_gate.is_closed()) {
+        /// Gate may already be closed if ::pause() had been called
+        co_await _gate.close();
+    }
+    if (_queue.size() > 0) {
+        vlog(
+          klog.debug,
+          "{} records were not pushed to the audit log before shutdown",
+          _queue.size());
+    }
+}
+
+ss::future<> audit_log_manager::pause() {
+    return container().invoke_on_all([](audit_log_manager& mgr) {
+        mgr._drain_timer.cancel();
+        return mgr._gate.close();
+    });
+}
+
+ss::future<> audit_log_manager::resume() {
+    return container().invoke_on_all([](audit_log_manager& mgr) {
+        if (mgr._gate.is_closed()) {
+            mgr._gate = ss::gate();
```

Comment:
does `pause` really need to wait on the gate to close in the sense that it isn't shutting down the service? it seems like you could pause/resume without that, and then not have the complexity of resetting the gate which is a bit of an anti pattern.

Reply:
I agree though, I have my reservations about this maybe there is another solution, will think about it

Reply:
It seems like it boils down to a decision about if you want pause to wait until everything stops. That's the way it seems to currently be working. But does it need to wait? Because you could introduce something like a semaphore with value 0 (when paused) or value int_max (when not paused). Probably other options too

Reply:
Oh that is a nice idea I think it would lend nicer to whats going on here, I can give that a shot